### PR TITLE
feat: Add extensions option allowing to load additional Jinja2 extensions

### PR DESCRIPTION
**Pull request by pawamoy**
Issue #208

### CODE REVIEW

File: copier/cli.py
Comment (pawamoy) [inline copier/cli.py:62]: Ah it's not needed here in fact. There will be no CLI option for this.

File: copier/main.py
Comment (pawamoy) [inline copier/main.py:392]: I chose to set default extensions here, so they are *always* loaded, whatever the template writer puts in `_extensions`. But maybe you think we should let writers override these?

File: copier/tools.py
Comment (pawamoy) [inline copier/tools.py:129]: Contrary to this function, the `to_nice_yaml` from Ansible filters does not seem to `rstrip()` the result. This is the reason the tests fail, but for the life of me I was not able to fix the test! I tried adding or removing newlines with multiple methods, but pytest would not have it. I'll continue trying of course, but if you want to to take a look I don't say no :sweat_smile:

Comment (Yajo) [reply copier/tools.py:129]: It's weird, but it doesn't seem like a big loss. You can modify the tests if you want.

Comment (pawamoy) [reply copier/tools.py:129]: Pffff, I was tired I guess, it's just that the same assert is repeated three times and I didn't get that until now... Tests now passing. I will add some more.

**Comment by pawamoy** [comment]
I've added tests, but I'm not sure I took the right approach. With this implementation:

1. default extensions (ansible, cookiecutter, etc.) are always loaded
2. extensions in copier.yml are also loaded (written as import strings)
3. extensions given through the `extensions` argument of `copier.copy` are also loaded

It means that a template *writer* can add extensions, or override the default ones, while a template *user* (given they run Copier programmatically, in Python) can add extensions (which seems useless?), or override the default or template ones.

I can see the use-case of point 3 (overriding a filter for example), but I'm not sure it makes sense. What are your thoughts?

**Comment by pawamoy** [comment]
Arf I messed up the rebase. Will fix it another day.

**Comment by Yajo** [comment]
> I can see the use-case of point 3 (overriding a filter for example), but I'm not sure it makes sense. What are your thoughts?

I think it's better to _not_ support this use case. It would produce confusion for template users and unexpected results for template designers.

This feature should exist *only* in the template (copier.yaml).

Thanks for your effort! I'll review deeper when you end the draft phase 😊

**Comment by Yajo** [comment]
> Arf I messed up the rebase. Will fix it another day.

I see some conflicts with current master. You might prefer to do a merge instead, which is usually easier. I can squash at merge later, so don't worry too much about git history right now (if that makes your life easier 😉 )

**Comment by pawamoy** [comment]
>  think it's better to not support this use case. It would produce confusion for template users and unexpected results for template designers.

Done! I removed the `extensions` parameter from `copier.copy`, templates users won't be able to mess with extensions :slightly_smiling_face: 

Side note: I have an unrelated test failing locally that I'm not sure how to fix? `test_jinja_defaults_v6_min_version`

```
copier.errors.UnknownCopierVersionWarning: Cannot check Copier version constraint.
```

**Comment by Yajo** [comment]
> Side note: I have an unrelated test failing locally that I'm not sure how to fix: `test_jinja_defaults_v6_min_version`

Maybe you need to pull with git tags.

**Comment by pawamoy** [comment]
This is ready for review. Maybe we could add more documentation? And a TODO for the `make_secret` as mentioned previously? The changelog also lags behind at v4, not sure if I should take care of it or if it's autogenerated. Waiting for your review (no rush)!

**Comment by Yajo** [comment]
> The changelog also lags behind at v4, not sure if I should take care of it or if it's autogenerated

huh... that's a problem I have to fix at some point. I'll open a separate issue, I need some strategy on that.

I'll take a deep review when I have some time, thank you!

File: copier/main.py
Comment (Yajo) [inline copier/main.py:381]: These 2 extensions are not needed, because `AnsibleCoreFiltersExtension` includes:

- `ans_random`
- `to_json`
- `to_nice_json`

File: copier/main.py
Comment (Yajo) [inline copier/main.py:382]: The point above makes me think about this one extension...

Cookiecutter is a big dependency. Since you're adding a way to easily include external extensions... Would it be acceptable to you to remove the whole cookiecutter thing?

After all, if you just `pipx inject copier cookiecutter` and add this extension to your template (or maybe [jinja2-slug](https://pypi.org/project/jinja2-slug/) if that's all you want), it should be enough for most cases.

File: copier/main.py
Comment (Yajo) [inline copier/main.py:383]: This is a convenient extension too, but again `AnsibleCoreFiltersExtension` would do the trick with `{{ '%Y-%m-%d %H:%M:%S' | strftime }}`.

So, in this case also, please remove the whole dependency.

I think we can live with just ansible-core-filters shipped by default.

---

😢 Sorry! I know it was me who asked you to include those dependencies... but after seeing the code in place, seeing how ansible-core-filters does the job, and also seeing how easy it would be to add extra extensions, I changed my mind. I'm very sorry! 🙏🏼

File: docs/configuring.md
Comment (Yajo) [inline docs/configuring.md:384]: So, this should disappear.

File: pyproject.toml
Comment (Yajo) [inline pyproject.toml:31]: This too should disappear.

File: tests/demo_extensions_additional/copier.yml
Comment (Yajo) [inline tests/demo_extensions_additional/copier.yml:1]: These days I'm more in favor of building the code tree in the test module. It seems cleaner and easier to reference the test with the template. I've done that for you.

File: docs/configuring.md
Comment (Yajo) [inline docs/configuring.md:368]: Do you think it makes sense to call this `jinja_extensions`? 🤔 

That maybe makes it more self-explanatory, and separates it from maybe copier extensions or other kind of extensions that might come in the future? Disclaimer: I'm not particularly thinking on doing that right now.

File: tests/test_updatediff.py
Comment (Yajo) [inline tests/test_updatediff.py:139]: I think all you really needed was to add the last `\n`, right? If so, please do that (to reduce the diff).

File: tests/test_updatediff.py
Comment (Yajo) [inline tests/test_updatediff.py:182]: Same about the diff.

File: tests/test_updatediff.py
Comment (Yajo) [inline tests/test_updatediff.py:211]: And here too.

**Review by Yajo — CHANGES_REQUESTED** [review]
I like very much how you developed this. Thank you! You'll have to revert some things though, see below (please don't hate me).

Also, some points that I couldn't comment inline:

- `now` and `make_secret` need to bubble a `FutureWarning` that indicates how these will disappear in a future major Copier release, and how to upgrade each of them. (`now` is commented below. `make_secret` can be replaced with `999999999999999999999999999999999|ans_random|hash("sha512")`)
- You need to document how to actually install more extensions. You can include some common ones in a small example catalog (in docs).
- The exception raised when an extension isn't found is not much helpful for template end users. Maybe that could be intercepted when creating the environment to display a more helpful error?

File: tests/test_updatediff.py
Comment (pawamoy) [reply tests/test_updatediff.py]: Sure, I deemed it was not that much noise to bring consistency on multiline-strings indentation. Feel free to do that in another commit/PR.

File: docs/configuring.md
Comment (pawamoy) [reply docs/configuring.md]: We already have `envops` and not `jinja_envops`, but yeah, `jinja_extensions` is fine, it's more explicit :slightly_smiling_face:

File: tests/demo_extensions_additional/copier.yml
Comment (pawamoy) [reply tests/demo_extensions_additional/copier.yml:1]: > I've done that for you.

Not sure to understand! I don't any new commit from you on this branch?

File: copier/main.py
Comment (pawamoy) [reply copier/main.py]: You're totally right, I was a bit troubled with adding such a dependency anyway!

File: copier/main.py
Comment (pawamoy) [reply copier/main.py]: Hahaha it's totally fine don't worry! :joy:

**Comment by pawamoy** [comment]
> I like very much how you developed this.

Well, your great refactor made this really easy!

I'll now work on the three additional points you mentioned :slightly_smiling_face:

**Comment by pawamoy** [comment]
It seems Ansible provides a `password_hash` filter that is not available in the extension: https://docs.ansible.com/ansible/2.3/playbooks_filters.html#hashing-filters. Too bad. I worry a bit that we won't be able to send PRs to this project as I can't find its sources on the web :confused:

**Comment by Yajo** [comment]
> I worry a bit that we won't be able to send PRs to this project as I can't find its sources on the web confused

They're not linked in pypi.org for some reason, but they are at https://gitlab.com/dreamer-labs/libraries/jinja2-ansible-filters/ as it seems.

File: tests/demo_extensions_additional/copier.yml
Comment (Yajo) [reply tests/demo_extensions_additional/copier.yml]: Ah! I forgot to upload it.

**Review by Yajo — COMMENTED** [review]
Just one little extra detail.

File: docs/configuring.md
Comment (pawamoy) [reply docs/configuring.md:381]: Huh, I seem to have starred this repo two weeks ago :O what am I doing X)
Thanks :)

**Comment by pawamoy** [comment]
Alright, just a bit of docs now, and we should be good.

**Comment by pawamoy** [comment]
Exception OK:

> Copier could not load some Jinja extensions:
No module named 'extension_not_installed'
Make sure to install these extensions alongside Copier itself.
See the docs at https://copier.readthedocs.io/en/latest/configuring/extensions

File: copier/main.py
Comment (Yajo) [inline copier/main.py:393]: Please use a specific exception class. See the errors module.

File: copier/main.py
Comment (Yajo) [inline copier/main.py:562]: And this? Were you unable to obtain it from elsewhere? Maybe the file system loader, or `_copier_conf.src_path`?

Comment (pawamoy) [reply copier/main.py:562]: Oops, forgot to remove that! Yes I was able to get that from the loader!

File: copier/main.py
Comment (pawamoy) [reply copier/main.py:393]: Do you mean something like this? 

```python
class ExtensionNotFoundError(UserMessageError, CopierError):
    """Extensions listed in the configuration could not be loaded."""
```

By the way, why do you inherit from both `UserMessageError` *and* `CopierError` when the first already inherits from the second?

Comment (Yajo) [reply copier/main.py:393]: > Do you mean something like this?

Yes



> By the way, why do you inherit from both `UserMessageError` _and_ `CopierError` when the first already inherits from the second?

Oops, that's a bug.

Comment (pawamoy) [reply copier/main.py:393]: I wrote like that anyway, for the sake of consistency. You'll be able to fix all error classes at once in a later commit/PR :slightly_smiling_face:

**Comment by pawamoy** [comment]
That's a lot of fixups, sorry :sweat_smile: 

I've fixed the last few bits like comments and docs. It's good on my end.

File: docs/configuring.md
Comment (Yajo) [inline docs/configuring.md:401]: I think it would be very helpful to list here cookiecutter extensions, because those won't appear while searching pypi, and would help people coming from cookiecutter to preserve some features they might be using.

File: docs/configuring.md
Comment (Yajo) [inline docs/configuring.md:405]: https://pypi.org/search/?q=jinja+extension would help too.

**Review by Yajo — COMMENTED** [review]
This looks extremely good, thanks for your valuable contribution.

There are a little couple of additions I'd like you to add to the docs before merging, please.

File: docs/configuring.md
Comment (pawamoy) [reply docs/configuring.md:401]: Totally. But maybe we could show how to do the same with Ansible filters, without a dependency to cookiecutter, since it's possible.

File: docs/configuring.md
Comment (pawamoy) [reply docs/configuring.md]: Done.

**Review by Yajo — APPROVED** [review]
Awesome!

**Comment by pawamoy** [comment]
Just wanted to say thank you again for taking the time to review this with me @Yajo! I really appreciate it 🙂

**Comment by Yajo** [comment]
I appreciate your time and effort too, thanks 😄

**Pull request merged**
